Improve error messaging on bad Pants version

NavneethJayendran opened the pull request:
Resolves #155

Handle the case where the Github API returns multiple results due to an inexact release tag match
by raising a new `TagNotFoundError` which also links the Pants releases page. Also wrap the typical
`CalledProcessError` with this new exception.

jsirois commented on tools/src/scie_pants/pants_version.py:
```diff
@@ -117,12 +120,22 @@ def determine_tag_version(
+            github_api_response = ptex.fetch_json(github_api_url, **headers)
+            # If the response is a list, multiple matches were found, which means the supplied tag
+            # is not an exact match against any tag
+            if isinstance(github_api_response, list):
```
It seems like it would make a but more sense to keep the comment mostly unchanged but change the test to `if not isinstance(github_api_response, dict)` since that is what is used / relied upon on line 138 below. As it stands, if the API returned a string a bool or a number, the same style of confusing exception as this fixes would be raised.

jsirois reviewed (commented):
Thanks @NavneethJayendran. Are you up for adding a test?

That might go here as a call to a new `test_bad_pants_version` test function you'd add lower down: https://github.com/pantsbuild/scie-pants/blob/b92a07685822a019c1682e23ca3ec100de07934d/package/src/test.rs#L98

NavneethJayendran commented:
@jsirois Sorry for the long response time. I adjusted the check to look for a dict specifically. @AIGeneratedUsername I added a test case for this error message as you suggested.

jsirois commented:
@NavneethJayendran thanks for picking this back up. It looks like there are test issue to work out still. I'm going to add a few folks as reviewers since I've stepped away from Pants-related development. They should be able to help you take this home or else recommend further reviewers.

kaos commented on package/src/test.rs:
```diff
@@ -38,7 +40,7 @@ fn decode_output(output: Vec<u8>) -> Result<String> {
 }
 
 fn assert_stderr_output(command: &mut Command, expected_messages: Vec<&str>) -> Output {
-    let output = execute(command.stderr(Stdio::piped())).unwrap();
+    let output = execute_no_error(command.stderr(Stdio::piped()));
```
I worry that this does no longer make the same assertion about the sub process' exit code as before for all the unrelated tests using this method now.

huonw replied:
As of #296, callers of `assert_stderr_output` can indicate whether the command should either succeed or fail.

thejcannon commented on tools/src/scie_pants/pants_version.py:
```diff
@@ -117,12 +121,22 @@ def determine_tag_version(
         github_api_url = (
             f"https://api.github.com/repos/pantsbuild/pants/git/refs/tags/{urllib.parse.quote(tag)}"
         )
+        github_releases_url = "https://github.com/pantsbuild/pants/releases"
```
So we're actually about to switch to a GitHub Release-based approach for any version >=2.

If you're able to sit a smidge, this code might look different when we make that switch.

NavneethJayendran replied:
Let me know when that happens and I can revisit.

huonw replied:
Thanks for waiting! The switch has happened now.

For version strings that look new (after `PANTS_PEX_GITHUB_RELEASE_VERSION`), this code path isn't used, as the function returns early. For older ones, definitely still worth considering improvements'; thanks!

For those new ones, Pants is then downloaded from GitHub directly, with some error handling about it not existing (but feel free to fine-tune if you have ideas!), in https://github.com/pantsbuild/scie-pants/blob/91477f6394e8eed048e093538b4a4c299514e44f/tools/src/scie_pants/install_pants.py#L88-L116

benjyw commented:
Hi @NavneethJayendran ! Just checking on the status of this?
